File: src/main/java/org/example/coin_laundry_app_backend/config/security/jwt/JWTAuthenticationFilter.java
package org.example.coin_laundry_app_backend.config.security.jwt;

import com.auth0.jwt.exceptions.JWTVerificationException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.web.filter.OncePerRequestFilter;

@RequiredArgsConstructor
public class JWTAuthenticationFilter extends OncePerRequestFilter {

    private static final String TOKEN_PREFIX = "Bearer ";
    private static final String EXCEPTION = "exception";

    private final JWTHelper jwtHelper;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
        FilterChain filterChain) throws ServletException, IOException {
        Optional.ofNullable(request.getHeader(HttpHeaders.AUTHORIZATION)).ifPresentOrElse(
            token -> {
                token = URLDecoder.decode(token, StandardCharsets.UTF_8);
                if (!token.contains(TOKEN_PREFIX)) {
                    request.setAttribute(EXCEPTION,
                        new JWTVerificationException("Invalid token type"));
                } else {
                    token = token.substring(7);
                    try {
                        var userId = jwtHelper.verify(token);
                        var authentication = new JWTAuthenticationToken(userId,
                            generateAuthority());
                        authentication.setDetails(
                            new WebAuthenticationDetailsSource().buildDetails(request)
                        );
                        SecurityContextHolder.getContext().setAuthentication(authentication);
                    } catch (Exception e) {
                        request.setAttribute(EXCEPTION, e);
                    }
                }
            },
            () -> request.setAttribute(EXCEPTION,
                new JWTVerificationException("Invalid token type"))
        );
        filterChain.doFilter(request, response);

    }

    private List<GrantedAuthority> generateAuthority() {
        return List.of(new SimpleGrantedAuthority("ROLE_USER"));
    }
}

File: carry-security/src/main/kotlin/com/carry/security/filter/JwtAuthenticationFilter.kt
package com.carry.security.filter

import com.carry.security.jwt.JwtProvider
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtAuthenticationFilter(
    private val jwtProvider: JwtProvider
) : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val token = resolveToken(request)
        if (token != null) {
            val principal = jwtProvider.parseToken(token)
            if (principal != null) {
                val authentication = UsernamePasswordAuthenticationToken(
                    principal.userId,
                    null,
                    listOf(SimpleGrantedAuthority("ROLE_${principal.role}"))
                )
                SecurityContextHolder.getContext().authentication = authentication
                MDC.put(MDC_USER_ID, principal.userId.toString())
            }
        }
        try {
            filterChain.doFilter(request, response)
        } finally {
            // 인증 성공 시 채운 userId를 항상 정리(스레드풀 재사용 누수 방지). 미설정 키 remove는 no-op.
            MDC.remove(MDC_USER_ID)
        }
    }

    private fun resolveToken(request: HttpServletRequest): String? {
        val bearer = request.getHeader("Authorization") ?: return null
        return if (bearer.startsWith("Bearer ")) bearer.substring(7) else null
    }

    private companion object {
        const val MDC_USER_ID = "userId"  // logback includeMdcKeyName과 일치
    }
}
